xscen/aggregate.py

```diff
@@ -467,3 +491,282 @@ def spatial_mean(
+      Size of the window over which the mean tempearture reaches wl and for over which
+      to calculate the climatological mean of the indicators.
+    min_window: int
+      Minimum number of years required for climatological mean to be computed.
+      If left at None, it will be window-2. This is to get a non-nan value for periods
+      that go from december to january (ex. DJF in QS-DEC indicators) because they will have
+      less occurrences than the window size because nans are dropped.
+    tas_base_period: list
+      Base period. The warming is calculated with respect to this period.
+    ignore_member: bool
+      Whether to use the row of the warming_level_csv, even if the member is wrong.
+    tas_csv: str
+      Path to a csv of annual global mean temperature.
+      If None, it will default to data/IPCC_annual_global_tas.csv which was built from
+      the IPCC atlas data ( https://github.com/IPCC-WG1/Atlas/tree/main/datasets-aggregated-regionally/data )
+      and extra data from pilot models of MRCC5 and ClimEx.
+    to_level:
+      The processing level to assign to the output.
+      Use "{wl}" in the string to dynamically include the warming level.
+
+    Returns
+    -------
+    xr.Dataset
+        Warming level dataset
+    """
+    if min_window is None:
+        min_window = window - 2
+
+    if tas_csv is None:
+        tas_csv = Path(__file__).parent / "data/IPCC_annual_global_tas.csv"
+
+    # get info on ds
+    id_ds = ds.attrs["cat:id"]
+    source_ds = ds.attrs["cat:source"]
+    exp_ds = ds.attrs["cat:experiment"]
+    member_ds = ds.attrs["cat:member"]
+    mip_era_ds = ds.attrs["cat:mip_era"]
+
+    info_ds = (
+        f"{mip_era_ds}_{source_ds}_{exp_ds}_*"
+        if ignore_member
+        else f"{mip_era_ds}_{source_ds}_{exp_ds}_{member_ds}"
+    )
+
+    # open csv
+    annual_tas = pd.read_csv(tas_csv, index_col="year")
+
+    # choose colum based in ds cat attrs
+    right_column = annual_tas.filter(regex=info_ds, axis=1)
+
+    if len(right_column.columns) > 1:
+        logger.info(
+            "More than one column of the csv was selected. Choosing the first one."
+        )
+        right_column = pd.DataFrame(right_column.iloc[:, 0])
+    elif len(right_column.columns) == 0:
+        # TODO: should this fail ?
+        logger.info(
+            f"No columns fit the 'cat:' attributes of the input dataset ({info_ds}). Returning None."
+        )
+        return None
+
+    logger.info(
+        f"Computing warming level +{wl}C for id: {id_ds} from column: {right_column.columns[0]}."
+    )
+
+    # compute reference temperature for the warming
+    mean_base = right_column.loc[tas_base_period[0] : tas_base_period[1]].mean()
+
+    yearly_diff = right_column - mean_base  # difference from reference
+    last_year = right_column.iloc[-1].name  # last available year
+
+    # get the star and end date of the window when the warming level is first reached
+    for year in np.arange(1850, 2101):
```

Comment:
What makes you think it is faster ?  (I don't know, this code from Chris) But , now I am thinking, it might be better to not hard code years.. Eventually, the data could go after 2100.

Reply:
changed it for a mix of both methods